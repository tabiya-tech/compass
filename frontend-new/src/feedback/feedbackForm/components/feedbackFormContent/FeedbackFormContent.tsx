import * as React from "react";
import { useState } from "react";
import { Divider, useMediaQuery } from "@mui/material";
import { Theme, useTheme } from "@mui/material/styles";
import MobileStepper from "@mui/material/MobileStepper";
import Box from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import PrimaryButton from "src/theme/PrimaryButton/PrimaryButton";
import stepsContent from "src/feedback/feedbackForm/stepsContent";
import StepsComponent from "src/feedback/feedbackForm/components/stepsComponent/StepsComponent";
import { PersistentStorageService } from "src/app/PersistentStorageService/PersistentStorageService";
import { FeedbackItem } from "src/feedback/feedbackForm/feedbackFormService/feedbackFormService.types";

interface FeedbackFormContentProps {
  notifySubmit: (formData: FeedbackItem[]) => void;
}

const uniqueId = "07d1ed9d-6ced-4a24-912a-9b89199df67f";

export const DATA_TEST_ID = {
  FEEDBACK_FORM_CONTENT: `feedback-form-content-${uniqueId}`,
  FEEDBACK_FORM_CONTENT_TITLE: `feedback-form-content-title-${uniqueId}`,
  FEEDBACK_FORM_CONTENT_QUESTIONS: `feedback-form-content-questions-${uniqueId}`,
  FEEDBACK_FORM_CONTENT_DIVIDER: `feedback-form-content-divider-${uniqueId}`,
  FEEDBACK_FORM_NEXT_BUTTON: `feedback-form-next-button-${uniqueId}`,
  FEEDBACK_FORM_BACK_BUTTON: `feedback-form-back-button-${uniqueId}`,
};

export const STORAGE = PersistentStorageService.storage;
export const FEEDBACK_FORM_ANSWERS_KEY = "userFeedbackResponses";

const FeedbackFormContent: React.FC<FeedbackFormContentProps> = ({ notifySubmit }) => {
  const theme = useTheme();
  const isSmallMobile = useMediaQuery((theme: Theme) => theme.breakpoints.down("sm"));
  const [activeStep, setActiveStep] = React.useState(0);
  const [answers, setAnswers] = useState<FeedbackItem[]>(() => {
    const savedAnswers = PersistentStorageService.getItem(STORAGE, FEEDBACK_FORM_ANSWERS_KEY);
    return savedAnswers ? JSON.parse(savedAnswers) : [];
  });

  const maxSteps = stepsContent.length;

  const handleNext = () => {
    if (activeStep === maxSteps - 1) {
      notifySubmit(answers);
      PersistentStorageService.removeItem(STORAGE, FEEDBACK_FORM_ANSWERS_KEY);
      setAnswers([]);
    } else {
      setActiveStep(prev => prev + 1);
    }
  };

  const handlePrevious = () => {
    setActiveStep((prevActiveStep) => prevActiveStep - 1);
  };

  const handleAnswerChange = (feedback: FeedbackItem) => {
    setAnswers((prevAnswers) => {
      const existingIndex = prevAnswers.findIndex((item) => item.question_id === feedback.question_id);

      let updatedAnswers;
      if (existingIndex !== -1) {
        updatedAnswers = [...prevAnswers];
        updatedAnswers[existingIndex] = feedback;
      } else {
        updatedAnswers = [...prevAnswers, feedback];
      }

      // Save updated answers to persistent storage
      PersistentStorageService.setItem(STORAGE, FEEDBACK_FORM_ANSWERS_KEY, JSON.stringify(updatedAnswers));

      return updatedAnswers;
    });
  };

  // Check if there is at least one answer
  const hasAnswers = Object.keys(answers).length > 0;

  return (
    <Box display="flex" flexDirection="column" gap={isSmallMobile ? 6 : 3} height="100%"
         data-testid={DATA_TEST_ID.FEEDBACK_FORM_CONTENT}>
      <Typography fontWeight="bold" color={theme.palette.text.secondary}
                  sx={{ fontSize: theme.typography.h6.fontSize }}
                  data-testid={DATA_TEST_ID.FEEDBACK_FORM_CONTENT_TITLE}
      >{stepsContent[activeStep].label}</Typography>
      <Box sx={{
        overflowY: "auto",
        flex: 1,
        display: "flex",
        flexDirection: "column",
        gap: isSmallMobile ? theme.spacing(6) : theme.spacing(3),
      }}
           data-testid={DATA_TEST_ID.FEEDBACK_FORM_CONTENT_QUESTIONS}
      >
        <StepsComponent questions={stepsContent[activeStep].questions} feedbackItems={answers} onChange={handleAnswerChange} />
      </Box>
      <Divider
        color="primary"
        sx={{ height: "0.2rem" }}
        data-testid={DATA_TEST_ID.FEEDBACK_FORM_CONTENT_DIVIDER}
      />
      <MobileStepper
        variant="dots"
        steps={maxSteps}
        position="static"
        activeStep={activeStep}
        nextButton={
          <PrimaryButton
            onClick={handleNext}
            disabled={activeStep === maxSteps - 1 && !hasAnswers}
            style={{ width: 100 }}
            data-testid={DATA_TEST_ID.FEEDBACK_FORM_NEXT_BUTTON}
          >
            {activeStep === maxSteps - 1 ? "Submit" : "Next"}
          </PrimaryButton>
        }
        backButton={
          <PrimaryButton
            variant="text"
            onClick={handlePrevious}
            disabled={activeStep === 0}
            style={{ color: theme.palette.common.black, width: 100 }}
            data-testid={DATA_TEST_ID.FEEDBACK_FORM_BACK_BUTTON}
          >
            Previous
          </PrimaryButton>
        }
        sx={{ height: "50px", paddingX: 0 }}
      />
    </Box>
  );
};

export default FeedbackFormContent;
