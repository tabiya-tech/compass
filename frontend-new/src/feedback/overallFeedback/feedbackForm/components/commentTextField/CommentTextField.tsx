import React from "react";
import { TextField, useTheme } from "@mui/material";

interface CommentTextFieldProps {
  placeholder?: string;
  value: string;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

const uniqueId = "ab74a074-67f6-4a8a-b2b2-a582d1d56c9c";

export const DATA_TEST_ID = {
  COMMENT_TEXT_FIELD: `comment-text-field-${uniqueId}`,
};

const CommentTextField = React.forwardRef<HTMLInputElement, CommentTextFieldProps>(
  ({ placeholder, value, onChange }, ref) => {
    const theme = useTheme();

    return (
      <TextField
        placeholder={placeholder}
        multiline
        fullWidth
        rows={3}
        variant="outlined"
        value={value}
        onChange={onChange}
        inputRef={ref}
        inputProps={{ "data-testid": DATA_TEST_ID.COMMENT_TEXT_FIELD }}
        sx={{ marginTop: theme.fixedSpacing(theme.tabiyaSpacing.xs) }}
      />
    );
  }
);

export default CommentTextField;
