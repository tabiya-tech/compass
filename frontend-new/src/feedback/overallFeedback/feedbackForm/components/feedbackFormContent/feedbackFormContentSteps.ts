import { useState, useEffect, useCallback } from "react";
import { useTranslation } from "react-i18next";
// REMOVED: import questions from "src/feedback/overallFeedback/feedbackForm/questions-en.json"; 
// The data will now be loaded dynamically inside the hook.
import {
    DetailedQuestion,
    QuestionType,
    YesNoEnum,
} from "src/feedback/overallFeedback/feedbackForm/feedbackForm.types";

interface Step {
    label: string;
    questions: DetailedQuestion[];
}

// Define the shape of your imported questions data for better type safety
interface QuestionsData {
    [key: string]: {
        question_text: string;
        comment_placeholder?: string;
        options?: {
            [key: string]: string;
        };
    };
}

/**
 * Custom hook to dynamically load and structure the feedback form steps
 * based on the current i18n locale.
 * @returns An object containing the structured steps and a loading state.
 */
export const useFeedbackFormContentSteps = () => {
    // We use `t` for translating static labels and `i18n` for locale management
    const { t, i18n } = useTranslation();
    const [questionsData, setQuestionsData] = useState<QuestionsData | null>(null);
    const [loading, setLoading] = useState(true);

    // Function to dynamically import the correct file
    const loadQuestions = useCallback(async (locale: string) => {
        setLoading(true);
        try {
            // Construct the file path dynamically. 
            // We use the full path relative to the project root/src for dynamic imports in Vite.
            const module = await import(/* @vite-ignore */ `src/feedback/overallFeedback/feedbackForm/questions-${locale}.json`);

            // JSON files often export as `default`.
            setQuestionsData(module.default || module);
        } catch (error) {
            console.error(`Failed to load questions for locale ${locale}.`, error);

            // Fallback: If the specific locale file fails, try to load a default (e.g., 'en-gb')
            if (locale !== 'en-gb') {
                console.info("Attempting to load fallback 'en-gb' locale.");
                loadQuestions('en-gb'); // Recursive call for fallback
                return;
            }
            // If fallback also fails, set to empty object
            console.error("Fallback 'en-gb' locale also failed to load.");
            setQuestionsData({});
        } finally {
            setLoading(false);
        }
    }, []);

    useEffect(() => {
        // Initial load based on current language
        loadQuestions(i18n.language);

        // Listen for language changes and reload questions
        i18n.on('languageChanged', loadQuestions);

        // Cleanup listener on component unmount
        return () => {
            i18n.off('languageChanged', loadQuestions);
        };
    }, [i18n, loadQuestions]);

    // Map the steps only if the questionsData is loaded
    const feedbackFormContentSteps: Step[] = questionsData
        ? [
            {
                // Translate the static step label using i18n's 't' function
                label: t("bias_experience_accuracy"), 
                questions: [
                    {
                        type: QuestionType.YesNo,
                        questionId: "perceived_bias",
                        // Use the dynamically loaded data
                        questionText: questionsData["perceived_bias"].question_text,
                        showCommentsOn: YesNoEnum.Yes,
                        placeholder: questionsData["perceived_bias"].comment_placeholder,
                    },
                    {
                        type: QuestionType.Checkbox,
                        questionId: "work_experience_accuracy",
                        questionText: questionsData["work_experience_accuracy"].question_text,
                        // Ensure options is not undefined before mapping
                        options: Object.entries(questionsData["work_experience_accuracy"].options || {}).map(([key, value]) => ({ key, value })),
                        // Translate static labels
                        lowRatingLabel: t("inaccurate"), 
                        highRatingLabel: t("very_accurate"),
                        placeholder: questionsData["work_experience_accuracy"].comment_placeholder,
                    },
                ],
            },
            {
                label: t("skill_accuracy"),
                questions: [
                    {
                        type: QuestionType.YesNo,
                        questionId: "clarity_of_skills",
                        questionText: questionsData["clarity_of_skills"].question_text,
                        showCommentsOn: YesNoEnum.No,
                        placeholder: questionsData["clarity_of_skills"].comment_placeholder,
                    },
                    {
                        type: QuestionType.YesNo,
                        questionId: "incorrect_skills",
                        questionText: questionsData["incorrect_skills"].question_text,
                        showCommentsOn: YesNoEnum.Yes,
                        placeholder: questionsData["incorrect_skills"].comment_placeholder,
                    },
                    {
                        type: QuestionType.YesNo,
                        questionId: "missing_skills",
                        questionText: questionsData["missing_skills"].question_text,
                        showCommentsOn: YesNoEnum.Yes,
                        placeholder: questionsData["missing_skills"].comment_placeholder,
                    },
                ],
            },
            {
                label: t("final_feedback"),
                questions: [
                    {
                        type: QuestionType.Rating,
                        questionId: "interaction_ease",
                        questionText: questionsData["interaction_ease"].question_text,
                        lowRatingLabel: t("difficult"),
                        highRatingLabel: t("easy"),
                        maxRating: 5,
                        placeholder: questionsData["interaction_ease"].comment_placeholder,
                    },
                    {
                        type: QuestionType.Rating,
                        questionId: "recommendation",
                        questionText: questionsData["recommendation"].question_text,
                        lowRatingLabel: t("unlikely"),
                        highRatingLabel: t("likely"),
                        maxRating: 5,
                    },
                    {
                        type: QuestionType.Rating,
                        questionId: "additional_feedback",
                        questionText: questionsData["additional_feedback"].question_text,
                        displayRating: false,
                        placeholder: questionsData["additional_feedback"].comment_placeholder,
                    },
                ],
            },
        ]
        : []; // Return empty array if data is still loading or failed to load

    return { feedbackFormContentSteps, loading };
};

export default useFeedbackFormContentSteps;