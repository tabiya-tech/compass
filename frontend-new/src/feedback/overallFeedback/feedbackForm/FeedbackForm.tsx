import React from "react";
import { Dialog, DialogContent, DialogTitle, Typography, useMediaQuery, useTheme } from "@mui/material";
import { Theme } from "@mui/material/styles";
import FeedbackFormContent from "src/feedback/overallFeedback/feedbackForm/components/feedbackFormContent/FeedbackFormContent";
import PrimaryIconButton from "src/theme/PrimaryIconButton/PrimaryIconButton";
import CloseIcon from "@mui/icons-material/Close";
import OverallFeedbackService from "src/feedback/overallFeedback/overallFeedbackService/OverallFeedback.service";
import { useSnackbar } from "src/theme/SnackbarProvider/SnackbarProvider";
import { FeedbackItem } from "src/feedback/overallFeedback/overallFeedbackService/OverallFeedback.service.types";
import { Backdrop } from "src/theme/Backdrop/Backdrop";
import UserPreferencesStateService from "src/userPreferences/UserPreferencesStateService";
import { FeedbackError } from "src/error/commonErrors";

export interface FeedbackFormProps {
  isOpen: boolean;
  notifyOnClose: (event: { name: CloseEventName }) => void;
  onFeedbackSubmit: () => void;
}

export enum CloseEventName {
  DISMISS = "DISMISS",
}

const uniqueId = "c6ba52ec-c1de-46ac-950b-f5354c6785ac";

export const DATA_TEST_ID = {
  FEEDBACK_FORM_DIALOG: `feedback-form-dialog-${uniqueId}`,
  FEEDBACK_FORM_DIALOG_TITLE: `feedback-form-dialog-title-${uniqueId}`,
  FEEDBACK_FORM_DIALOG_BUTTON: `feedback-form-dialog-button-${uniqueId}`,
  FEEDBACK_FORM_DIALOG_ICON_BUTTON: `feedback-form-dialog-icon-button-${uniqueId}`,
  FEEDBACK_FORM_DIALOG_CONTENT: `feedback-form-dialog-content-${uniqueId}`,
};

const FeedbackForm: React.FC<FeedbackFormProps> = ({ isOpen, notifyOnClose, onFeedbackSubmit }) => {
  const theme = useTheme();
  const { enqueueSnackbar } = useSnackbar();
  const isSmallMobile = useMediaQuery((theme: Theme) => theme.breakpoints.down("sm"));
  const [isSubmitting, setIsSubmitting] = React.useState(false);

  const handleClose = () => {
    notifyOnClose({ name: CloseEventName.DISMISS });
  };

  const handleFeedbackSubmit = async (formData: FeedbackItem[]): Promise<void> => {
    setIsSubmitting(true);
    handleClose();
    try {
      const userPreferences = UserPreferencesStateService.getInstance().getUserPreferences();
      if (!userPreferences?.sessions.length) {
        throw new Error("User has no sessions");
      }

      const sessionId = userPreferences.sessions[0];
      const overallFeedbackService = new OverallFeedbackService(sessionId);
      await overallFeedbackService.sendFeedback(formData);

      enqueueSnackbar("Feedback submitted successfully!", { variant: "success" });
      onFeedbackSubmit();
    } catch (error) {
      console.error(new FeedbackError("Failed to submit feedback", error as Error));
      enqueueSnackbar("Failed to submit feedback. Please try again later.", { variant: "error" });
    } finally {
      setIsSubmitting(false);
    }
  };

  return (
    <>
      <Dialog
        open={isOpen}
        onClose={handleClose}
        maxWidth="sm"
        fullWidth={true}
        fullScreen={isSmallMobile}
        data-testid={DATA_TEST_ID.FEEDBACK_FORM_DIALOG}
        sx={{
          "& .MuiDialog-paper": {
            height: "100%",
            display: "flex",
            flexDirection: "column",
            gap: theme.spacing(4),
          },
        }}
      >
        <DialogTitle component="div" sx={{ display: "flex", alignItems: "center", justifyContent: "space-between" }}>
          <Typography variant="h3" data-testid={DATA_TEST_ID.FEEDBACK_FORM_DIALOG_TITLE}>
            Help us improve!
          </Typography>
          <PrimaryIconButton
            onClick={handleClose}
            title="close feedback form"
            sx={{ color: theme.palette.text.secondary }}
            data-testid={DATA_TEST_ID.FEEDBACK_FORM_DIALOG_BUTTON}
          >
            <CloseIcon data-testid={DATA_TEST_ID.FEEDBACK_FORM_DIALOG_ICON_BUTTON} />
          </PrimaryIconButton>
        </DialogTitle>
        <DialogContent data-testid={DATA_TEST_ID.FEEDBACK_FORM_DIALOG_CONTENT}>
          <FeedbackFormContent notifySubmit={handleFeedbackSubmit} />
        </DialogContent>
      </Dialog>
      <Backdrop isShown={isSubmitting} message="Submitting feedback..." />
    </>
  );
};

export default FeedbackForm;
