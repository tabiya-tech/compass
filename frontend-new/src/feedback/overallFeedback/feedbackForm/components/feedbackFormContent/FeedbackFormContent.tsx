import * as React from "react";
import { useState } from "react";
import { Divider, useMediaQuery } from "@mui/material";
import { Theme, useTheme } from "@mui/material/styles";
import MobileStepper from "@mui/material/MobileStepper";
import Box from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import PrimaryButton from "src/theme/PrimaryButton/PrimaryButton";
import feedbackFormContentSteps from "src/feedback/overallFeedback/feedbackForm/components/feedbackFormContent/feedbackFormContentSteps";
import StepsComponent from "src/feedback/overallFeedback/feedbackForm/components/stepsComponent/StepsComponent";
import { PersistentStorageService } from "src/app/PersistentStorageService/PersistentStorageService";
import { FeedbackItem } from "src/feedback/overallFeedback/overallFeedbackService/OverallFeedback.service.types";
import SecondaryButton from "src/theme/SecondaryButton/SecondaryButton";

interface FeedbackFormContentProps {
  notifySubmit: (formData: FeedbackItem[]) => void;
}

const uniqueId = "07d1ed9d-6ced-4a24-912a-9b89199df67f";

export const DATA_TEST_ID = {
  FEEDBACK_FORM_CONTENT: `feedback-form-content-${uniqueId}`,
  FEEDBACK_FORM_CONTENT_TITLE: `feedback-form-content-title-${uniqueId}`,
  FEEDBACK_FORM_CONTENT_QUESTIONS: `feedback-form-content-questions-${uniqueId}`,
  FEEDBACK_FORM_CONTENT_DIVIDER: `feedback-form-content-divider-${uniqueId}`,
  FEEDBACK_FORM_NEXT_BUTTON: `feedback-form-next-button-${uniqueId}`,
  FEEDBACK_FORM_BACK_BUTTON: `feedback-form-back-button-${uniqueId}`,
};

const FeedbackFormContent: React.FC<FeedbackFormContentProps> = ({ notifySubmit }) => {
  const theme = useTheme();
  const isSmallMobile = useMediaQuery((theme: Theme) => theme.breakpoints.down("sm"));
  const [activeStep, setActiveStep] = React.useState(0);
  const [answers, setAnswers] = useState<FeedbackItem[]>(() => {
    return PersistentStorageService.getOverallFeedback();
  });

  const maxSteps = feedbackFormContentSteps.length;

  const handleNext = () => {
    if (activeStep === maxSteps - 1) {
      notifySubmit(answers);
      PersistentStorageService.clearOverallFeedback();
      setAnswers([]);
    } else {
      setActiveStep((prev) => prev + 1);
    }
  };

  const handlePrevious = () => {
    setActiveStep((prevActiveStep) => prevActiveStep - 1);
  };

  const handleAnswerChange = (feedback: FeedbackItem) => {
    setAnswers((prevAnswers) => {
      const existingIndex = prevAnswers.findIndex((item) => item.question_id === feedback.question_id);

      let updatedAnswers;
      if (existingIndex !== -1) {
        updatedAnswers = [...prevAnswers];
        updatedAnswers[existingIndex] = feedback;
      } else {
        updatedAnswers = [...prevAnswers, feedback];
      }

      // Save updated answers to persistent storage
      PersistentStorageService.setOverallFeedback(updatedAnswers);

      return updatedAnswers;
    });
  };

  // Check if there is at least one answer
  const hasAnswers = Object.keys(answers).length > 0;

  return (
    <Box
      display="flex"
      flexDirection="column"
      gap={isSmallMobile ? 6 : 3}
      height="100%"
      data-testid={DATA_TEST_ID.FEEDBACK_FORM_CONTENT}
    >
      <Typography
        fontWeight="bold"
        color={theme.palette.text.secondary}
        sx={{ fontSize: theme.typography.h6.fontSize }}
        data-testid={DATA_TEST_ID.FEEDBACK_FORM_CONTENT_TITLE}
      >
        {feedbackFormContentSteps[activeStep].label}
      </Typography>
      <Box
        sx={{
          overflowY: "auto",
          flex: 1,
          display: "flex",
          flexDirection: "column",
          gap: isSmallMobile ? theme.spacing(6) : theme.spacing(3),
        }}
        data-testid={DATA_TEST_ID.FEEDBACK_FORM_CONTENT_QUESTIONS}
      >
        <StepsComponent
          questions={feedbackFormContentSteps[activeStep].questions}
          feedbackItems={answers}
          onChange={handleAnswerChange}
        />
      </Box>
      <Divider color="primary" sx={{ height: "0.2rem" }} data-testid={DATA_TEST_ID.FEEDBACK_FORM_CONTENT_DIVIDER} />
      <MobileStepper
        variant="dots"
        steps={maxSteps}
        position="static"
        activeStep={activeStep}
        nextButton={
          <PrimaryButton
            onClick={handleNext}
            disabled={activeStep === maxSteps - 1 && !hasAnswers}
            style={{ width: 100 }}
            data-testid={DATA_TEST_ID.FEEDBACK_FORM_NEXT_BUTTON}
          >
            {activeStep === maxSteps - 1 ? "Submit" : "Next"}
          </PrimaryButton>
        }
        backButton={
          <SecondaryButton
            onClick={handlePrevious}
            disabled={activeStep === 0}
            data-testid={DATA_TEST_ID.FEEDBACK_FORM_BACK_BUTTON}
          >
            Previous
          </SecondaryButton>
        }
        sx={{ height: "50px", paddingX: 0 }}
      />
    </Box>
  );
};

export default FeedbackFormContent;
