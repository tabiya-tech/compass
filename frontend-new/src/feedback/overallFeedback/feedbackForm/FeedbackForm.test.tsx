// mute the console
import "src/_test_utilities/consoleMock";

import FeedbackForm, { DATA_TEST_ID, FeedbackCloseEvent } from "src/feedback/overallFeedback/feedbackForm/FeedbackForm";
import { render, screen } from "src/_test_utilities/test-utils";
import { fireEvent, waitFor } from "@testing-library/react";
import { DATA_TEST_ID as FEEDBACK_FORM_CONTENT_DATA_TEST_ID } from "src/feedback/overallFeedback/feedbackForm/components/feedbackFormContent/FeedbackFormContent";
import UserPreferencesStateService from "src/userPreferences/UserPreferencesStateService";
import { useSnackbar } from "src/theme/SnackbarProvider/SnackbarProvider";
import {
  SensitivePersonalDataRequirement,
  Language,
} from "src/userPreferences/UserPreferencesService/userPreferences.types";
import OverallFeedbackService from "src/feedback/overallFeedback/overallFeedbackService/OverallFeedback.service";
import feedbackFormContentSteps from "src/feedback/overallFeedback/feedbackForm/components/feedbackFormContent/feedbackFormContentSteps";
import { FeedbackError } from "src/error/commonErrors";
import { DATA_TEST_ID as COMMENT_TEXT_FIELD_TEST_ID } from "src/feedback/overallFeedback/feedbackForm/components/commentTextField/CommentTextField";

// mock the feedback service
jest.mock("src/feedback/overallFeedback/overallFeedbackService/OverallFeedback.service");

// mock the snackbar provider
jest.mock("src/theme/SnackbarProvider/SnackbarProvider", () => {
  const actual = jest.requireActual("src/theme/SnackbarProvider/SnackbarProvider");
  return {
    ...actual,
    __esModule: true,
    useSnackbar: jest.fn().mockReturnValue({
      enqueueSnackbar: jest.fn(),
      closeSnackbar: jest.fn(),
    }),
  };
});

// mock the user preferences state service
jest.mock("src/userPreferences/UserPreferencesStateService", () => ({
  getInstance: jest.fn().mockReturnValue({
    getUserPreferences: jest.fn().mockReturnValue({
      sessions: [1234],
    }),
  }),
}));

describe("FeedbackForm", () => {
  test("should render component successfully", () => {
    // GIVEN the component
    const givenFeedbackForm = <FeedbackForm isOpen={true} notifyOnClose={jest.fn()} />;

    // WHEN the component is rendered
    render(givenFeedbackForm);

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND the feedback form dialog to be in the document
    const feedbackFormContainer = screen.getByTestId(DATA_TEST_ID.FEEDBACK_FORM_DIALOG);
    expect(feedbackFormContainer).toBeInTheDocument();
    // AND the feedback form dialog title to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.FEEDBACK_FORM_DIALOG_TITLE)).toBeInTheDocument();
    // AND the feedback form dialog button to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.FEEDBACK_FORM_DIALOG_BUTTON)).toBeInTheDocument();
    // AND the feedback form dialog icon button to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.FEEDBACK_FORM_DIALOG_ICON_BUTTON)).toBeInTheDocument();
    // AND the feedback form dialog content to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.FEEDBACK_FORM_DIALOG_CONTENT)).toBeInTheDocument();
    // AND to match the snapshot
    expect(feedbackFormContainer).toMatchSnapshot();
  });

  test("should call handleClose when close button is clicked", () => {
    // GIVEN the component
    const mockHandleClose = jest.fn();
    const givenFeedbackForm = (
      <FeedbackForm isOpen={true} notifyOnClose={mockHandleClose} />
    );
    // AND the component is rendered
    render(givenFeedbackForm);

    // WHEN the close button is clicked
    const closeButton = screen.getByTestId(DATA_TEST_ID.FEEDBACK_FORM_DIALOG_ICON_BUTTON);
    fireEvent.click(closeButton);

    // THEN expect the notifyOnClose function to have been called
    expect(mockHandleClose).toHaveBeenCalled();
  });

  describe("FeedbackForm handleFeedbackSubmit", () => {
    const mockSendFeedback = jest.fn();

    beforeEach(() => {
      jest.clearAllMocks();
      (OverallFeedbackService as jest.Mocked<typeof OverallFeedbackService>).prototype.sendFeedback = mockSendFeedback;
    });

    test("should call handleFeedbackSubmit when submit button is clicked", async () => {
      // GIVEN the component
      const mockHandleClose = jest.fn();
      const givenFeedbackForm = (
        <FeedbackForm isOpen={true} notifyOnClose={mockHandleClose} />
      );
      // AND the component is rendered
      render(givenFeedbackForm);
      // AND there is at least one answer
      const input = screen.getAllByTestId(COMMENT_TEXT_FIELD_TEST_ID.COMMENT_TEXT_FIELD);
      fireEvent.change(input[0], { target: { value: "This is a comment" } });

      // WHEN the submit button is clicked
      const nextButton = screen.getByTestId(FEEDBACK_FORM_CONTENT_DATA_TEST_ID.FEEDBACK_FORM_NEXT_BUTTON);
      for (let i = 0; i < feedbackFormContentSteps.length; i++) {
        fireEvent.click(nextButton);
      }

      // THEN expect the notifyOnClose to have been called
      await waitFor(() => expect(mockHandleClose).toHaveBeenCalledWith(FeedbackCloseEvent.SUBMIT));
      // AND the sendFeedback function to have been called
      expect(mockSendFeedback).toHaveBeenCalled();
      // AND the snackbar to have been called
      await waitFor(() =>
        expect(useSnackbar().enqueueSnackbar).toHaveBeenCalledWith("Feedback submitted successfully!", {
          variant: "success",
        })
      );
    });

    test("should call handleFeedbackSubmit when submit button is clicked and handle error", async () => {
      (console.error as jest.Mock).mockClear();

      // GIVEN getUserPreferences returns a user without any session
      jest.spyOn(UserPreferencesStateService.getInstance(), "getUserPreferences").mockReturnValue({
        accepted_tc: new Date(),
        user_id: "0001",
        language: Language.en,
        sessions: [],
        sessions_with_feedback: [],
        has_sensitive_personal_data: false,
        sensitive_personal_data_requirement: SensitivePersonalDataRequirement.NOT_REQUIRED,
      });

      // WHEN the component is rendered
      render(<FeedbackForm isOpen={true} notifyOnClose={jest.fn()} />);
      // AND there is at least one answer
      const input = screen.getAllByTestId(COMMENT_TEXT_FIELD_TEST_ID.COMMENT_TEXT_FIELD);
      fireEvent.change(input[0], { target: { value: "This is a comment" } });
      // AND the submit button is clicked
      const nextButton = screen.getByTestId(FEEDBACK_FORM_CONTENT_DATA_TEST_ID.FEEDBACK_FORM_NEXT_BUTTON);
      for (let i = 0; i < feedbackFormContentSteps.length; i++) {
        fireEvent.click(nextButton);
      }

      // THEN expect an error message to be shown
      await waitFor(() => {
        expect(console.error).toHaveBeenCalledWith(
          new FeedbackError("Failed to submit feedback", new Error("User has no sessions"))
        );
      });
      // AND the snackbar to have been called
      await waitFor(() =>
        expect(useSnackbar().enqueueSnackbar).toHaveBeenCalledWith(
          "Failed to submit feedback. Please try again later.",
          { variant: "error" }
        )
      );
    });
  });
});
