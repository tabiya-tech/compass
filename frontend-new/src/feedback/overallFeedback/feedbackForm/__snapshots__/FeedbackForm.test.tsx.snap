// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackForm should render component successfully 1`] = `
<div
  class="MuiDialog-root MuiModal-root css-1hldcek-MuiModal-root-MuiDialog-root"
  data-testid="feedback-form-dialog-c6ba52ec-c1de-46ac-950b-f5354c6785ac"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-testid="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
    role="presentation"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    tabindex="-1"
  >
    <div
      aria-labelledby=":r0:"
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth css-dmoqe9-MuiPaper-root-MuiDialog-paper"
      role="dialog"
    >
      <div
        class="MuiTypography-root MuiTypography-h2 MuiDialogTitle-root css-1l7du2y-MuiTypography-root-MuiDialogTitle-root"
        id=":r0:"
      >
        <h3
          class="MuiTypography-root MuiTypography-h3 css-xcrvgq-MuiTypography-root"
          data-testid="feedback-form-dialog-title-c6ba52ec-c1de-46ac-950b-f5354c6785ac"
        >
          Help us improve!
        </h3>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-14ql4g7-MuiButtonBase-root-MuiIconButton-root"
          data-testid="feedback-form-dialog-button-c6ba52ec-c1de-46ac-950b-f5354c6785ac"
          tabindex="0"
          title="close feedback form"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-etbowg-MuiSvgIcon-root"
            data-testid="feedback-form-dialog-icon-button-c6ba52ec-c1de-46ac-950b-f5354c6785ac"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiDialogContent-root css-h86owe-MuiDialogContent-root"
        data-testid="feedback-form-dialog-content-c6ba52ec-c1de-46ac-950b-f5354c6785ac"
      >
        <div
          data-testid="feedback-form-content-07d1ed9d-6ced-4a24-912a-9b89199df67f"
        />
      </div>
    </div>
  </div>
  <div
    data-testid="sentinelEnd"
    tabindex="0"
  />
</div>
`;
